src/promise/pTuple.ts

```diff
@@ -0,0 +1,17 @@
+/**
+ *
+ * Wraps async calls in try catch blocks
+ * to simplify syntax.
+ *
+ * source: https://github.com/scopsy/await-to-js/blob/master/src/await-to-js.ts
+ *
+ */
+export async function pTuple<T, U = Error>(
```

Comment:
Good.

Great that you followed my advice to remove `errorExt`.

What about my suggestion/comment that in 99% of the cases the `U` type is not needed and plain `Error` suffices?

I just found one more argument to support my suggestion. Even if we have a custom Error type, there's no 100% guarantee that if the error is thrown that it would be of that type and not from some other place with the generic `Error` type. Like, an error that you don't anticipate. E.g, look at the default typings in TypeScript when you do `try {} catch (err)`. The `err` type is actually `any` (even more generic than `Error`), cause there's no guarantee of what error will be thrown. And in cases (probably less than 1% of times) where we want to assume some specific error type - using `(err as SomeError).someProperty` will be just fine.

Reply:
Actually, I do use error casting.  When you deal with custom errors is not only about checking the error type (test) but working with the error data. Some errors have custom properties that you won't "type" if we use Error any.

This is less useful on the backend where probably you just let the errors throw, but In the frontend, you want to properly handle each error type. For instance:
```
const [errorSigningIn] = await pTo<void, HttpError>(
      this.authService.socialAuth(auth, privacyConsent),
    )
    if (errorSigningIn && errorSigningIn.data.code === ErrorCode.ACCOUNT_EXISTS) { 
     // do something
  }
```
If we don't cast to HttpError, then we won't have access to data.code as the default Error object doesn't provide it.

![image](https://user-images.githubusercontent.com/48744933/112306721-ddce8800-8c7e-11eb-87af-b9e2f01ae8e7.png)

What do you think?

Reply:
I would write the equivalent code like this:

```ts
const [errorSigningIn] = await pTo(this.authService.socialAuth(auth, privacyConsent))
if ((errorSigningIn as HttpError)?.data.code === ErrorCode.ACCOUNT_EXISTS) { 
   // do something
}
```

But I see that it's subjective, not strictly-better.

What confuses me the most with your snippet is that you need to provide both types `<void, HttpError>`, while `void` is otherwise perfectly inferred by typescript.

But again, I agree that it's subjective.

I'll approve and let you decide 👍 

(or, I will be very curious to hear what other developers think/prefer)

Reply:
@stinaa @duryno @ryangibbsnc do you have an opinion?:)

Reply:
BTW, back to the "async stacktraces" topic. Have you tested that this code keeps the stacktrace and doesn't require you to add extra-await like this?

```ts
const [errorSigningIn] = await pTo(this.authService.socialAuth(auth, privacyConsent))
// vs
const [errorSigningIn] = await pTo(await this.authService.socialAuth(auth, privacyConsent))
```

Reply:
> What confuses me the most with your snippet is that you need to provide both types <void, HttpError>, while void is otherwise perfectly inferred by typescript.

I'm leaning a bit more toward `(errorSigningIn as HttpError)?.data.code` because of this ☝️, but it's not a strong opinion; i.e., I'm not against `U = Error` + `pTo<void, HttpError>`

Reply:
I think if we can do less `something as Something` that would be great! 😄

Reply:
> you need to provide both types <void, HttpError>
In this case you need void, just to cast the return data of "socialAuth", in this case, is void.

> BTW, back to the "async stacktraces" topic. Have you tested that this code keeps the stacktrace and doesn't require you to add extra-await like this?

If you use the double await the promise will throw without catch, as it would be resolved inside the run function and not inside pTuple. So we need to use only 1 await
![image](https://user-images.githubusercontent.com/48744933/112447750-f0fe5800-8d51-11eb-936d-15b5120c3a2b.png)

Funny that in this time using return await foo() and just returning foo() gives us the same full stack trace

![image](https://user-images.githubusercontent.com/48744933/112448053-463a6980-8d52-11eb-91aa-f214ae932f9f.png)
![image](https://user-images.githubusercontent.com/48744933/112448165-69fdaf80-8d52-11eb-8702-2e8277eddd1a.png)

Should I merge this PR now?

Reply:
> Should I merge this PR now?

Go ahead